pkg/customresourcestate/registry_factory.go

```diff
@@ -257,22 +244,54 @@ func (c *compiledGauge) Values(v interface{}) (result []eachValue, errs []error)
 
 type compiledInfo struct {
 	compiledCommon
+	labelFromKey string
 }
 
 func (c *compiledInfo) Values(v interface{}) (result []eachValue, errs []error) {
-	if vs, isArray := v.([]interface{}); isArray {
-		for _, obj := range vs {
+	onError := func(err ...error) {
+		errs = append(errs, fmt.Errorf("%s: %v", c.Path(), err))
+	}
+
+	switch iter := v.(type) {
+	case []interface{}:
+		for _, obj := range iter {
 			ev, err := c.values(obj)
 			if len(err) > 0 {
-				errs = append(errs, err...)
+				onError(err...)
 				continue
 			}
 			result = append(result, ev...)
 		}
-		return
+	default:
+		value, err := c.values(v)
+		if err != nil {
+			onError(err...)
+			break
+		}
+		// labelFromKey logic
+		if vv, ok := v.(map[string]interface{}); ok {
+			for key, val := range vv {
+				if key != "" && c.labelFromKey != "" {
+					n, err := toFloat64(val, false)
+					if err != nil {
+						onError(err)
+						continue
+					}
+					result = append(result, eachValue{
+						Labels: map[string]string{
+							c.labelFromKey: key,
+						},
+						Value: n,
+					})
+				}
+			}
+		}
+		if len(result) == 0 {
+			result = value
+		}
```

Comment:
Could we perhaps split that into two different cases, one for `map[string]interface{}` and the other for `default`, having it embedded into the default case confused me at first, I would prefer a logic similar to the one with have today with the gauge as it is easier to understand.